src/dns/query.cpp

```diff
@@ -3,46 +3,306 @@
+                Answer answer;
+                answer.code = code;
+                answer.ttl = ttl;
+                if (family == PF_INET) {
+                    answer.ipv4 = string;
+                    answer.type = QueryTypeId::A;
+                } else if (family == PF_INET6) {
+                    answer.ipv6 = string;
+                    answer.type = QueryTypeId::AAAA;
+                }
+                logger->info("dns - adding '%s'", string);
+                answers.push_back(answer);
+            }
+
+        } else {
+            logger->warn("dns - too many addresses");
+            code = DNS_ERR_UNKNOWN;
+        }
+
+    } else {
+        logger->warn("dns - invalid response type");
+        code = DNS_ERR_UNKNOWN;
     }
+    return answers;
+}
+
+static void handle_resolve(int code, char type, int count, int ttl,
+                    void *addresses, void *opaque) {
+
+    auto context = static_cast<QueryContext *>(opaque);
+
+    context->message.error_code = code;
+
+    switch (code) {
+        case DNS_ERR_NONE:
+        case DNS_ERR_FORMAT:
+        case DNS_ERR_SERVERFAILED:
+        case DNS_ERR_NOTEXIST:
+        case DNS_ERR_NOTIMPL:
+        case DNS_ERR_REFUSED:
+        case DNS_ERR_TRUNCATED:
+        case DNS_ERR_NODATA:
+            context->message.rtt = mk::time_now() - context->ticks;
+            break;
+        default:
+            context->message.rtt = 0.0;
+            break;
+    }
+
+    context->message.answers = build_answers_evdns(code, type, count, ttl,
+            addresses);
+    if (context->message.error_code != DNS_ERR_NONE) {
+        context->callback(mk::dns::dns_error(context->message.error_code), context->message);
+    } else {
+        context->callback(NoError(), context->message);
+    }
+
+    delete context;
+}
+
+void query(QueryClass dns_class, QueryType dns_type, std::string name,
+            std::function<void(Error, Message)> cb,
+            Settings settings,
+            Poller *poller) {
+
+    Message message;
+    Query query;
+
+    // XXX add proper exception handling for evbase creation
+    evdns_base *base = create_evdns_base(settings, poller);
+
+    if (dns_class != QueryClassId::IN) {
+        cb(UnsupportedClassError(), nullptr);
+        return;
+    }
+
+    // Allow PTR queries
+    if (dns_type == QueryTypeId::PTR) {
+        std::string s;
+        if ((s = mk::unreverse_ipv4(name)) != "") {
+            dns_type = QueryTypeId::REVERSE_A;
+            name = s;
+        } else if ((s = mk::unreverse_ipv6(name)) != "") {
+            dns_type = QueryTypeId::REVERSE_AAAA;
+            name = s;
+        } else {
+            cb(InvalidNameForPTRError(), nullptr);
+            return;
+        }
+    }
+
+    query.type = dns_type;
+    query.qclass = dns_class;
+    query.name = name;
+    message.queries.push_back(query);
+
+    //
+    // We explain above why we don't store the return value
+    // of the evdns_base_resolve_xxx() functions below
```

Comment:
This comment references a comment that was lost in the refactoring. I'd reinstate the original comment, which was:

```
    //
    // Note: evdns_base_resolve_xxx() return a evdns_request
    // object that may be used to cancel the request. Yet, the
    // semantic of cancelling a request is such that evdns
    // could invoke the callback of a pending request if such
    // request was cancelled when its corresponding response
    // was already pending. This seems to be confirmed by
    // comments in libevent's evdns.c:
    //
    //     This does nothing if the request's callback is
    //     already running (pending_cb is set)
    //
    // For the above reason this class does not explicitly
    // cancel pending evdns requests and uses the `cancelled`
    // variable to keep track of cancelled requests.
    //
```